lib/virtual-fs/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+        Self {
+            opener,
+            conf: OpenOptionsConfig {
+                read: false,
+                write: false,
+                create_new: false,
+                create: false,
+                append: false,
+                truncate: false,
+            },
+        }
+    }
+
+    pub fn read(&mut self, read: bool) -> &mut Self {
+        self.conf.read = read;
+        self
+    }
+
+    pub fn write(&mut self, write: bool) -> &mut Self {
+        self.conf.write = write;
+        self
+    }
+
+    pub fn append(&mut self, append: bool) -> &mut Self {
+        self.conf.append = append;
+        self
+    }
+
+    pub fn truncate(&mut self, truncate: bool) -> &mut Self {
+        self.conf.truncate = truncate;
+        self
+    }
+
+    pub fn create(&mut self, create: bool) -> &mut Self {
+        self.conf.create = create;
+        self
+    }
+
+    pub fn create_new(&mut self, create_new: bool) -> &mut Self {
+        self.conf.create_new = create_new;
+        self
+    }
+
+    pub fn open<P: AsRef<Path>>(&mut self, path: P) -> Result<Box<dyn VirtualFile>, FsError> {
+        self.opener.open(path.as_ref(), &self.conf)
+    }
+}
+
+/// This trait relies on your file closing when it goes out of scope via `Drop`
+#[typetag::serde(tag = "type")]
+pub trait VirtualFile: fmt::Debug + Send + Write + Read + Seek + 'static + Upcastable {
+    /// the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn last_accessed(&self) -> u64;
+
+    /// the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn last_modified(&self) -> u64;
+
+    /// the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn created_time(&self) -> u64;
+
+    /// set the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn set_last_accessed(&self, _last_accessed: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_accessed due to using the default implementation", self);
+    }
+
+    /// set the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn set_last_modified(&self, _last_modified: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_modified due to using the default implementation", self);
+    }
+
+    /// set the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn set_created_time(&self, _created_time: u64) {
+        debug!(
+            "{:?} did nothing in VirtualFile::set_created_time to using the default implementation",
+            self
+        );
+    }
+
+    /// the size of the file in bytes
+    fn size(&self) -> u64;
+
+    /// Change the size of the file, if the `new_size` is greater than the current size
+    /// the extra bytes will be allocated and zeroed
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError>;
+
+    /// Request deletion of the file
+    fn unlink(&mut self) -> Result<(), FsError>;
```

Comment:
Can we have this as `self` (so it's consumed)?

Reply:
Thinking more, I think perhaps `unlink` and `rename` can be in the `FileSystem` api rather than in the `VirtualFile` itself. It makes things a bit easier to reason about I think

Reply:
Yeah, it's due to WasiFile that it is this way, I agree that it's probably possible to move some of that logic away, it's really just a clean up for the file itself. I think `self` is problematic in trait objects, `Drop::drop` takes self by `&mut` so  I think this is fine

Reply:
Ok, I added a side comment on other place... perhaps we can delete this method and use only the filesystem one (`remove_file`)

Reply:
Yeah, I'm in no rush to do this yet (as I haven't done it) because I think depending on how we want to compose different types of VirtualFiles and file systems it may still make sense to have this as-is.

Reply:
Working out the exact lines between the abstractions is still a work in progress